### ISSUE TITLE
Proposed Answer for Issue #2

### ISSUE THREAD

**Issue by whyzhow** [My-Mujica/LSTM-Optuna#3]
Subject: Regarding Optuna Pruning implementation in LSTM training
     Hi there! Thank you for your question. Integrating Optuna's pruning feature is one of the best ways to speed up hyperparameter optimization for LSTM models. Here is a breakdown of how it works in the context of this project:
## 1. Manual Pruning vs. Lightning Callback
If you are using Vanilla PyTorch (as seen in this repository) rather than PyTorch Lightning, the pruning is typically handled manually within the training loop. The PyTorchLightningPruningCallback is specifically designed for the Lightning Trainer and won't work directly with a standard Python for epoch in range(...) loop.
## 2. How to implement it correctly
To ensure unpromising trials are terminated early, you need to incorporate two key steps inside your epoch loop:
• trial.report(value, step): This sends the current performance (e.g., Validation Loss) to Optuna.
• trial.should_prune(): This checks if the current trial's performance is significantly worse than previous successful trials.
## 3. Code Example
Here is how the logic should look inside your objective function:
def objective(trial):

```
    # ... model setup ...
    
    for epoch in range(n_epochs):
        train(model, optimizer)  # Your training function
        val_loss = validate(model)  # Your validation function
        
        # 1. Report the intermediate result
        trial.report(val_loss, epoch)
        
        # 2. Handle pruning (terminate the trial if it's performing poorly)
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()
            
    return val_loss
```
## 4. Important Note on Samplers
For pruning to take effect, you must use a Pruner (like MedianPruner or HyperbandPruner) when creating your study:

```
study = optuna.create_study(
    direction="minimize",
    pruner=optuna.pruners.MedianPruner() # Enables the pruning logic
)

```
Summary: In this project, the pruning is handled manually. By raising optuna.exceptions.TrialPruned(), Optuna stops the current training execution and moves on to the next set of hyperparameters, saving you significant computational time.
